File: R Files/NFL Field.R
#####################################################################################
#####################################################################################
## This script produces a ggplot version of a regulation NFL field. Each unit      ##
## in x and y is equivalent to one foot (12 in) and all parts of the model are     ##
## drawn to scale.                                                                 ##
#####################################################################################
#####################################################################################

library(ggplot2)
library(png)
library(grid)
library(raster)
library(magick)

draw_nfl = function(){
  nfl_logo = readPNG('League Logos/nfl.png')
  nfl_logo = rasterGrob(nfl_logo, interpolate = TRUE)
  
  logo_borders = data.frame(xmin = -18, xmax = 18, ymin = -30, ymax = 30)
  
  field = data.frame(xmin = -190, xmax = 190, ymin = -90, ymax = 90)
  sideline = data.frame(xmin = c(-180, -180), xmax = c(180, 180), ymin = c(-86, 80), ymax = c(-80, 86))
  endline = data.frame(xmin = c(-186, 180), xmax = c(-180, 186), ymin = c(-86, -86), ymax = c(86, 86))
  goalline = data.frame(xmin = c(-150 - (2/12), 150 - (2/12)), xmax = c(-150 + (2/12), 150 + (2/12)), ymin = c(-86, -86), ymax = c(86, 86))
  midline = data.frame(xmin = -2/12, xmax = 2/12, ymin = -80 + 4/12, ymax = 80 - 4/12)
  yd_line = data.frame(xmin = -147 - (2/12), xmax = -147 + (2/12), ymin = -80 + (4/12), ymax = -78 + (4/12))
  
  minor_yd_lines = data.frame(xmin = rep(NA, 49), xmax = rep(NA, 49), ymin = rep(NA, 49), ymax = rep(NA, 49))
  
  for(i in 1:49){
    minor_yd_lines$xmin[i] = (0 - (3 * i)) - (2/12)
    minor_yd_lines$xmax[i] = (0 - (3 * i)) + (2/12)
    minor_yd_lines$ymin[i] = -80 + (4/12)
    minor_yd_lines$ymax[i] = -78 + (4/12)
  }
  
  minor_yd_lines = rbind(minor_yd_lines, data.frame(xmin = rep(NA, 49), xmax = rep(NA, 49), ymin = rep(NA, 49), ymax = rep(NA, 49)))
  
  for(i in 50:98){
    minor_yd_lines$xmin[i] = (0 - (3 * (i - 49))) - (2/12)
    minor_yd_lines$xmax[i] = (0 - (3 * (i - 49))) + (2/12)
    minor_yd_lines$ymin[i] = 78 - (4/12)
    minor_yd_lines$ymax[i] = 80 - (4/12)
  }
  
  minor_yd_lines = rbind(minor_yd_lines, data.frame(xmin = rep(NA, 49), xmax = rep(NA, 49), ymin = rep(NA, 49), ymax = rep(NA, 49)))
  
  for(i in 99:147){
    minor_yd_lines$xmin[i] = (0 - (3 * (i - 98))) - (2/12)
    minor_yd_lines$xmax[i] = (0 - (3 * (i - 98))) + (2/12)
    minor_yd_lines$ymin[i] = -22 + (4/12)
    minor_yd_lines$ymax[i] = -20 + (4/12)
  }
  
  minor_yd_lines = rbind(minor_yd_lines, data.frame(xmin = rep(NA, 49), xmax = rep(NA, 49), ymin = rep(NA, 49), ymax = rep(NA, 49)))
  
  for(i in 148:196){
    minor_yd_lines$xmin[i] = (0 - (3 * (i - 147))) - (2/12)
    minor_yd_lines$xmax[i] = (0 - (3 * (i - 147))) + (2/12)
    minor_yd_lines$ymin[i] = 20 - (4/12)
    minor_yd_lines$ymax[i] = 22 - (4/12)
  }
  
  minor_yd_lines = rbind(minor_yd_lines, data.frame(xmin = -minor_yd_lines$xmax, xmax = -minor_yd_lines$xmin, ymin = minor_yd_lines$ymin, ymax = minor_yd_lines$ymax))
  
  major_yd_lines = data.frame(
    xmin = seq(0, 150, by = 15) - (2/12),
    xmax = seq(0, 150, by = 15) + (2/12),
    ymin = -80 + (4/12),
    ymax = 80 - (4/12)
  )
  
  major_yd_lines = rbind(major_yd_lines, data.frame(xmin = -major_yd_lines$xmax, xmax = -major_yd_lines$xmin, ymin = major_yd_lines$ymin, ymax = major_yd_lines$ymax))
  
  hashes = data.frame(xmin = rep(NA, 9), xmax = rep(NA, 9), ymin = rep(NA, 9), ymax = rep(NA, 9))
  for(i in 1:9){
    hashes$xmin[i] = (0 - (15 * i)) - (10/12)
    hashes$xmax[i] = (0 - (15 * i)) + (10/12)
    hashes$ymin[i] = -20 + (4/12)
    hashes$ymax[i] = -20 + (2/12)
  }
  
  hashes = rbind(
    hashes,
    data.frame(xmin = hashes$xmin, xmax = hashes$xmax, ymin = -hashes$ymax, ymax = -hashes$ymin)
  )
  
  hashes = rbind(
    hashes,
    data.frame(xmin = -hashes$xmax, xmax = -hashes$xmin, ymin = hashes$ymin, ymax = hashes$ymax)
  )
  
  extra_pt_mark = data.frame(xmin = c(-141 - (2/12), 141 - (2/12)), xmax = c(-141 + (2/12), 141 + (2/12)), ymin = c(-1, -1), ymax = c(1, 1))
  
  arrows_40 = data.frame(
    x = c(
      -36 - (6/12), -36 - (6/12), -36 - ((sqrt((36^2) - 36))/12), -36 - (6/12), -36 - (6/12),
      -36 - (6/12), -36 - (6/12), -36 - ((sqrt((36^2) - 36))/12), -36 - (6/12), -36 - (6/12)
    ),
    y = c(
      -44, -44 + (9/12), -44, -44 - (9/12), -44,
      44, 44 - (9/12), 44, 44 + (9/12), 44
    )
  )
  
  arrows_30 = data.frame(
    x = c(-66 - (6/12), -66 - (6/12), -66 - ((sqrt((36^2) - 36))/12), -66 - (6/12), -66 - (6/12),
          -66 - (6/12), -66 - (6/12), -66 - ((sqrt((36^2) - 36))/12), -66 - (6/12), -66 - (6/12)
    ),
    
    y = c( -44, -44 + (9/12), -44, -44 - (9/12), -44,
           44, 44 - (9/12), 44, 44 + (9/12), 44
    )
  )
  
  arrows_20 = data.frame(
    x = c(
      -96 - (6/12), -96 - (6/12), -96 - ((sqrt((36^2) - 36))/12), -96 - (6/12), -96 - (6/12),
      -96 - (6/12), -96 - (6/12), -96 - ((sqrt((36^2) - 36))/12), -96 - (6/12), -96 - (6/12)
    ),
    y = c(
      -44, -44 + (9/12), -44, -44 - (9/12), -44,
      44, 44 - (9/12), 44, 44 + (9/12), 44
    )
  )
  
  arrows_10 = data.frame(
    x = c(
      -126 - (6/12), -126 - (6/12), -126 - ((sqrt((36^2) - 36))/12), -126 - (6/12), -126 - (6/12),
      -126 - (6/12), -126 - (6/12), -126 - ((sqrt((36^2) - 36))/12), -126 - (6/12), -126 - (6/12)
    ),
    y = c(
      -44, -44 + (9/12), -44, -44 - (9/12), -44,
      44, 44 - (9/12), 44, 44 + (9/12), 44
    )
  )
  
  ggplot() +
    theme_void() +
    theme(
      panel.border = element_blank(),
      panel.background = element_blank(),
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks = element_blank()
    ) +
    coord_fixed() +
    geom_rect(data = field, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#196f0c') +
    geom_rect(data = sideline, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    geom_rect(data = endline, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    geom_rect(data = goalline, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    geom_rect(data = minor_yd_lines, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    geom_rect(data = major_yd_lines, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    geom_rect(data = extra_pt_mark, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff') +
    annotation_custom(nfl_logo, xmin = logo_borders$xmin, xmax = logo_borders$xmax, ymin = logo_borders$ymin, ymax = logo_borders$ymax) +
    annotate('text', label = '5', x = -3, y = -44, col = '#ffffff') +
    annotate('text', label = '0', x = c(3, -27, 33, -57, 63, -87, 93, -117, 123), y = -44, col = '#ffffff') +
    annotate('text', label = '4', x = c(-33, 27), y = -44, col = '#ffffff') +
    annotate('text', label = '3', x = c(-63, 57), y = -44, col = '#ffffff') +
    annotate('text', label = '2', x = c(-93, 87), y = -44, col = '#ffffff') +
    annotate('text', label = '1', x = c(-123, 117), y = -44, col = '#ffffff') +
    annotate('text', label = '5', x = 3, y = 44, col = '#ffffff', angle = 180) +
    annotate('text', label = '0', x = c(-3, 27, -33, 57, -63, 87, -93, 117, -123), y = 44, col = '#ffffff', angle = 180) +
    annotate('text', label = '4', x = c(33, -27), y = 44, col = '#ffffff', angle = 180) +
    annotate('text', label = '3', x = c(63, -57), y = 44, col = '#ffffff', angle = 180) +
    annotate('text', label = '2', x = c(93, -87), y = 44, col = '#ffffff', angle = 180) +
    annotate('text', label = '1', x = c(123, -117), y = 44, col = '#ffffff', angle = 180) +
    geom_polygon(data = arrows_40, aes(x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_30, aes(x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_20, aes(x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_10, aes(x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_40, aes(-x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_30, aes(-x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_20, aes(-x, y), fill = '#ffffff') +
    geom_polygon(data = arrows_10, aes(-x, y), fill = '#ffffff') +
    geom_rect(data = hashes, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = '#ffffff')
}

File: Python Files/NFL Field.py
#####################################################################################
#####################################################################################
## This script produces a matplotlib version of a regulation NFL field. Each unit  ##
## in x and y is equivalent to one foot (12 in) and all parts of the model are     ##
## drawn to scale.                                                                 ##
#####################################################################################
#####################################################################################

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Create field, sideline markings, goal lines, and the 50 yard line
field = pd.DataFrame({
    'x':[-190, 190, 190, -190, -190],
    'y':[-90, -90, 90, 90, -90]
})

sideline = pd.DataFrame({
    'x':[-180, 180, 180, -180, -180],
    'y':[-86, -86, -80, -80, -86]
}).append(
    pd.DataFrame({
        'x':[-180, 180, 180, -180, -180],
        'y':[80, 80, 86, 86, 80]
    })
)

endline = pd.DataFrame({
    'x':[-186, -180, -180, -186, -186],
    'y':[-86, -86, 86, 86, -86]
}).append(
    pd.DataFrame({
        'x':[180, 186, 186, 180, 180],
        'y':[-86, -86, 86, 86, -86]
    })
)

goal_line = pd.DataFrame({
    'x':[-150 - (2/12), -150 + (2/12), -150 + (2/12), -150 - (2/12), -150 - (2/12)],
    'y':[-86, -86, 86, 86, -86]
}).append(
    pd.DataFrame({
        'x':[150 - (2/12), 150 + (2/12), 150 + (2/12), 150 - (2/12), 150 - (2/12)],
        'y':[-86, -86, 86, 86, -86]
    })
)

midline = pd.DataFrame({
    'x':[-2/12, 2/12, 2/12, -2/12, -2/12],
    'y':[-80 + (4/12), -80 + (4/12), -20 + (4/12), -20 + (4/12), -80 + (4/12)]
}).append(
    pd.DataFrame({
        'x':[-2/12, 2/12, 2/12, -2/12, -2/12],
        'y':[20 - (4/12), 20 - (4/12), 80 - (4/12), 80 - (4/12), 20 - (4/12)]
    })
)

# Create all of the minor yard lines (there are four sets)
minor_yd_lines_b = pd.DataFrame({
    'x':[],
    'y':[]
})

for i in range(1, 50):
    minor_yd_lines_b = minor_yd_lines_b.append(
        pd.DataFrame({
            'x':[(0 - (3 * i)) - (2/12), (0 - (3 * i)) + (2/12), (0 - (3 * i)) + (2/12), (0 - (3 * i)) - (2/12), (0 - (3 * i)) - (2/12)],
            'y':[-80 + (4/12), -80 + (4/12), -78 + (4/12), -78 + (4/12), -80 + (4/12)]
        })
    )

minor_yd_lines_b = minor_yd_lines_b.append(
    pd.DataFrame({
        'x':-1 * minor_yd_lines_b['x'],
        'y':minor_yd_lines_b['y']
    })
)

minor_yd_lines_t = pd.DataFrame({
    'x':minor_yd_lines_b['x'],
    'y':-1 * minor_yd_lines_b['y']
})

minor_yd_lines_l = pd.DataFrame({
    'x':minor_yd_lines_b['x'],
    'y':minor_yd_lines_b['y'] + 58
})

minor_yd_lines_u = pd.DataFrame({
    'x':minor_yd_lines_l['x'],
    'y':-1 * minor_yd_lines_l['y']
})

# Create the major yard lines (every 5 yards excluding the 50)
major_yd_lines = pd.DataFrame({
    'x':[],
    'y':[]
})

major_yd_lines_ft = np.arange(15, 150, 15)

for i in range(0, len(major_yd_lines_ft)):
    major_yd_lines = major_yd_lines.append(
        pd.DataFrame({
            'x':[major_yd_lines_ft[i] - (2/12), major_yd_lines_ft[i] + (2/12), major_yd_lines_ft[i] + (2/12), major_yd_lines_ft[i] - (2/12), major_yd_lines_ft[i] - (2/12)],
            'y':[-80 + (4/12), -80 + (4/12), 80 - (4/12), 80 - (4/12), -80 + (4/12)]
        })
    )
    
major_yd_lines = major_yd_lines.append(
    pd.DataFrame({
        'x':-1 * major_yd_lines['x'],
        'y':major_yd_lines['y']
    })
)

# Create the hash marks at every 5 yard line
hashes = pd.DataFrame({
    'x':[],
    'y':[]
})

for i in range(1, 10):
    hashes = hashes.append(
        pd.DataFrame({
            'x':[(0 - (15 * i)) - (10/12), (0 - (15 * i)) + (10/12), (0 - (15 * i)) + (10/12), (0 - (15 * i)) - (10/12), (0 - (15 * i)) - (10/12)],
            'y':[-20 + (4/12), -20 + (4/12), -20 + (2/12), -20 + (2/12), -20 + (4/12)]
        })
    )
    
hashes_l = hashes.append(
    pd.DataFrame({
        'x':-1 * hashes['x'],
        'y':hashes['y']
    })
)

hashes_u = pd.DataFrame({
    'x':hashes_l['x'],
    'y':hashes_l['y'] + 39.5
})

# Create the extra point marker
extra_pt_mark = pd.DataFrame({
    'x':[-141 - (2/12), -141 + (2/12), -141 + (2/12), -141 - (2/12), -141 - (2/12)],
    'y':[-1, -1, 1, 1, -1]
})

extra_pt_mark = extra_pt_mark.append(
    pd.DataFrame({
        'x':-1 * extra_pt_mark['x'],
        'y':extra_pt_mark['y']
    })
)

# Create the arrows next to each of the numbers on the field indicating direction
arrow_40 = pd.DataFrame({
    'x':[-36.5 - (6/12), -36.5 - (6/12), -36.5 - ((np.sqrt((36 ** 2) - 36))/12), -36.5 - (6/12), -36.5 - (6/12)],
    'y':[-44, -44 + (9/12), -44, -44 - (9/12), -44]
})

arrow_40_l = arrow_40.append(
    pd.DataFrame({
        'x':-1 * arrow_40['x'],
        'y':arrow_40['y']
    })
)

arrow_40_u = pd.DataFrame({
    'x':arrow_40_l['x'],
    'y':arrow_40_l['y'] + 92
})

arrow_30 = pd.DataFrame({
    'x':[-66.5 - (6/12), -66.5 - (6/12), -66.5 - ((np.sqrt((36 ** 2) - 36))/12), -66.5 - (6/12), -66.5 - (6/12)],
    'y':[-44, -44 + (9/12), -44, -44 - (9/12), -44]
})

arrow_30_l = arrow_30.append(
    pd.DataFrame({
        'x':-1 * arrow_30['x'],
        'y':arrow_30['y']
    })
)

arrow_30_u = pd.DataFrame({
    'x':arrow_30_l['x'],
    'y':arrow_30_l['y'] + 92
})

arrow_20 = pd.DataFrame({
    'x':[-96.5 - (6/12), -96.5 - (6/12), -96.5 - ((np.sqrt((36 ** 2) - 36))/12), -96.5 - (6/12), -96.5 - (6/12)],
    'y':[-44, -44 + (9/12), -44, -44 - (9/12), -44]
})

arrow_20_l = arrow_20.append(
    pd.DataFrame({
        'x':-1 * arrow_20['x'],
        'y':arrow_20['y']
    })
)

arrow_20_u = pd.DataFrame({
    'x':arrow_20_l['x'],
    'y':arrow_20_l['y'] + 92
})

arrow_10 = pd.DataFrame({
    'x':[-126.5 - (6/12), -126.5 - (6/12), -126.5 - ((np.sqrt((36 ** 2) - 36))/12), -126.5 - (6/12), -126.5 - (6/12)],
    'y':[-44, -44 + (9/12), -44, -44 - (9/12), -44]
})

arrow_10_l = arrow_10.append(
    pd.DataFrame({
        'x':-1 * arrow_10['x'],
        'y':arrow_10['y']
    })
)

arrow_10_u = pd.DataFrame({
    'x':arrow_10_l['x'],
    'y':arrow_10_l['y'] + 92
})

#################
# Make the plot #
#################
fig, ax = plt.subplots()

ax.set_aspect('equal')
fig.set_size_inches(50, 22.2)
ax.xaxis.set_visible(False)
ax.yaxis.set_visible(False)

# Set field color
ax.set_facecolor('#196f0c')

# Put NFL logo at midfield
img = 'League Logos/nfl.png'
img = plt.imread(img)
ax.imshow(img, extent = [-18., 18., -18., 18.])

# Add sidelines, goal line, and 50 yard line
ax.fill(sideline['x'], sideline['y'], '#ffffff')
ax.fill(endline['x'], endline['y'], '#ffffff')
ax.fill(goal_line['x'], goal_line['y'], '#ffffff')
ax.fill(midline['x'], midline['y'], '#ffffff')

# Add minor yard lines and major yard lines
ax.fill(minor_yd_lines_b['x'], minor_yd_lines_b['y'], '#ffffff')
ax.fill(minor_yd_lines_t['x'], minor_yd_lines_t['y'], '#ffffff')
ax.fill(minor_yd_lines_l['x'], minor_yd_lines_l['y'], '#ffffff')
ax.fill(minor_yd_lines_u['x'], minor_yd_lines_u['y'], '#ffffff')
ax.fill(major_yd_lines['x'], major_yd_lines['y'], '#ffffff')

# Add hash marks and extra point markers
ax.fill(hashes_l['x'], hashes_l['y'], '#ffffff')
ax.fill(hashes_u['x'], hashes_u['y'], '#ffffff')
ax.fill(extra_pt_mark['x'], extra_pt_mark['y'], '#ffffff')

# Add the numbers to the field
ax.text(-6.5, -46, '5', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(.25, 46, '5', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(1.25, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-6.75, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(-36.5, -46, '4', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-29.75, 46, '4', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(-28.75, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-36.75, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(-66.5, -46, '3', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-59.75, 46, '3', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(-58.75, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-66.75, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(-96.5, -46, '2', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-89.75, 46, '2', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(-88.75, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-96.75, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(-126.5, -46, '1', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-119.75, 46, '1', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(-118.75, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(-126.75, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(23.5, -46, '4', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(30.25, 46, '4', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(31.25, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(23.25, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(53.5, -46, '3', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(60.25, 46, '3', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(61.25, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(53.25, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(83.5, -46, '2', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(90.25, 46, '2', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(91.25, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(83.25, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

ax.text(113.5, -46, '1', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(120.25, 46, '1', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)
ax.text(121.25, -46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold')
ax.text(113.25, 46, '0', color = '#ffffff', fontsize = 50, fontweight = 'bold', rotation = 180)

# Add the arrows to the field
ax.fill(arrow_40_l['x'], arrow_40_l['y'], '#ffffff')
ax.fill(arrow_40_u['x'], arrow_40_u['y'], '#ffffff')
ax.fill(arrow_30_l['x'], arrow_30_l['y'], '#ffffff')
ax.fill(arrow_30_u['x'], arrow_30_u['y'], '#ffffff')
ax.fill(arrow_20_l['x'], arrow_20_l['y'], '#ffffff')
ax.fill(arrow_20_u['x'], arrow_20_u['y'], '#ffffff')
ax.fill(arrow_10_l['x'], arrow_10_l['y'], '#ffffff')
ax.fill(arrow_10_u['x'], arrow_10_u['y'], '#ffffff')

plt.show()